README.md
# RLicor
Read Licor files

Install remotely with the following line:
```
devtools::install_github("erikerhardt/RLicor")
```


NEWS.md
---
output:
  md_document:
    variant: markdown_github
---

# RLicor 0.0.00

6/11/2019
Initialized package

# RLicor 0.0.01

6/12/2019
Functions for reading Licor 6400 and 6800



R/read_Licor_6400.R
#' read Licor 6400 file
#'
#' Read Licor file
#' Find the row where the column headers and data begin.
#' Separate any remarks into its own column.
#'
#' @param Licor_fn Licor filename to read
#'
#' @return Licor A list including a \code{header} list and \code{data} tibble
#' @importFrom magrittr %>%
#' @importFrom readr read_file
#' @importFrom stringr str_split str_replace_na
#' @importFrom purrr map flatten_chr safely
#' @importFrom dplyr bind_rows mutate select everything
#' @importFrom lubridate ymd_hms
#' @importFrom stats na.omit
#'
#' @export
#'
read_Licor_6400 <-
  function (Licor_fn) {

  # header and data, separated
  Licor <- list()

  Licor$model <- 6400

  # concepts from http://www.ericrscott.com/2018/01/17/li-cor-wrangling/

  # read data in as text to remove header and remark rows from column headers and data
  Licor_raw <- readr::read_file(Licor_fn)

  header_pattern  <- "\"OPEN \\d\\.\\d\\.\\d"
  data_pattern    <- "\\$STARTOFDATA\\$"

  # splits into individual bouts
  Licor_bouts <- stringr::str_split(Licor_raw, header_pattern, simplify = TRUE)

  # splits further to separate headers from actual data
  Licor_header_data <- stringr::str_split(Licor_bouts, data_pattern, simplify = FALSE)

  # separate header and data, remove empty elements (first is always empty)

  ## Header
  Licor$header <-
    Licor_header_data %>%
    purrr::map(`[`, 1) %>% #equivalent to doing raw_split2[[i]][2] for every element "i"
    purrr::flatten_chr() #converts to a vector
  Licor$header <-
    Licor$header[!(Licor$header == "")]
  Licor$header <-
    stringr::str_split(
      Licor$header
    , "\n"
    , simplify = TRUE
    ) %>%
    t()

  # date
  Licor$datetime <-
    lubridate::mdy_hms(Licor$header) %>%
    stats::na.omit()



  ## Data
  dat_temp <-
    Licor_header_data %>%
    purrr::map(`[`, 2) %>% # equivalent to doing Licor_header_data[[i]][2] for every element "i"
    purrr::flatten_chr() # converts to a vector
  dat_temp <-
    dat_temp[!is.na(dat_temp)]

  dat_temp2 <-
    dat_temp %>%
    purrr::map(readr::read_tsv, skip = 1) %>%
    dplyr::bind_rows()

  # create a "safe" version of as.integer() that returns a list of a result and error
  # returns error for text remarks, returns value for integer observation numbers
  safe_as.int <-
    purrr::safely(as.integer)

  dat_temp3 <-
    dat_temp2 %>%
    dplyr::mutate(
      # create a comment column to indicate if an "Obs" is actually a remark
      comment = is.na(safe_as.int(Obs)$result)
      # copy those remarks to the remark column
    , remark = ifelse(comment == TRUE, Obs, NA)
      # remove remarks from Obs column
    , Obs = ifelse(comment == FALSE, Obs, NA)
    ) %>%
    # move the remark column the the begining
    dplyr::select(
      remark
    , dplyr::everything()
    ) %>%
    # remove the temporary comment column.
    dplyr::select(-comment) %>%
    dplyr::mutate(
      Obs = as.numeric(Obs)
    )
    #%>%
    ## replace NA with the literal string "NA" so str_* functions from stringr can deal with it
    #dplyr::mutate(
    #  remark = stringr::str_replace_na(remark)
    #)

  Licor$data <-
    dat_temp3

  return(Licor)
}




R/read_Licor_6800.R
#' read Licor 6800 file
#'
#' Read Licor file
#' Find the row where the column headers and data begin.
#' Separate any remarks into its own column.
#'
#' @param Licor_fn Licor filename to read
#'
#' @return Licor A list including a \code{header} list and \code{data} tibble
#' @importFrom magrittr %>%
#' @importFrom readr read_file
#' @importFrom stringr str_split str_replace_na
#' @importFrom purrr map flatten_chr safely
#' @importFrom dplyr bind_rows mutate select everything
#' @importFrom lubridate ymd_hms
#' @importFrom stats na.omit
#'
#' @export
#'
read_Licor_6800 <-
  function (Licor_fn) {

  # header and data, separated
  Licor <- list()

  Licor$model <- 6800

  # concepts from http://www.ericrscott.com/2018/01/17/li-cor-wrangling/

  # read data in as text to remove header and remark rows from column headers and data
  Licor_raw <- readr::read_file(Licor_fn)

  header_pattern  <- "[Header]"
  data_pattern    <- "[Data]"

  # splits into individual bouts
  Licor_bouts <- stringr::str_split(Licor_raw, stringr::fixed(header_pattern), simplify = TRUE)

  # splits further to separate headers from actual data
  Licor_header_data <- stringr::str_split(Licor_bouts, stringr::fixed(data_pattern), simplify = FALSE)

  # separate header and data, remove empty elements (first is always empty)

  ## Header
  Licor$header <-
    Licor_header_data %>%
    purrr::map(`[`, 1) %>% #equivalent to doing raw_split2[[i]][2] for every element "i"
    purrr::flatten_chr() #converts to a vector
  Licor$header <-
    Licor$header[!(Licor$header == "")]

  # date
  Licor$header <-
    stringr::str_split(
      Licor$header
    , "\n"
    , simplify = TRUE
    ) %>%
    t()

  # date
  Licor$datetime <-
    lubridate::ymd_hms(Licor$header) %>%
    stats::na.omit()


  ## Data
  dat_temp <-
    Licor_header_data %>%
    purrr::map(`[`, 2) %>% # equivalent to doing Licor_header_data[[i]][2] for every element "i"
    purrr::flatten_chr() # converts to a vector
  dat_temp <-
    dat_temp[!is.na(dat_temp)]

  # Because the header has 3 lines
  #
  read_tsv_skip_before_after_header <- function(dat, skip_pre = 2, skip_post = 1) {
    # if header is on row 2 and data starts on row 4,
    # then skip_pre = 1 and skip_post = 1
    #source_names  <- names(readr::read_tsv(dat, skip = skip_pre - 1, n_max = 0))
    col_names     <- names(readr::read_tsv(dat, skip = skip_pre    , n_max = 0))
    #unit_names    <- names(readr::read_tsv(dat, skip = skip_pre + 1, n_max = 0))
    dat2 <- readr::read_tsv(dat, skip = skip_pre + 1 + skip_post, col_names = col_names)
    return(dat2)
  }

  dat_temp2 <-
    dat_temp %>%
    purrr::map(read_tsv_skip_before_after_header) %>%
    dplyr::bind_rows()

  # create a "safe" version of as.integer() that returns a list of a result and error
  # returns error for text remarks, returns value for integer observation numbers
  safe_as.int <-
    purrr::safely(as.integer)

  dat_temp3 <-
    dat_temp2 %>%
    dplyr::mutate(
      # create a comment column to indicate if an "obs" is actually a remark
      comment = is.na(safe_as.int(obs)$result)
      # copy those remarks to the remark column
    , remark = ifelse(comment == TRUE, obs, NA)
      # remove remarks from obs column
    , obs = ifelse(comment == FALSE, obs, NA)
    ) %>%
    # move the remark column the the begining
    dplyr::select(
      remark
    , dplyr::everything()
    ) %>%
    # remove the temporary comment column.
    dplyr::select(-comment) %>%
    dplyr::mutate(
      obs = as.numeric(obs)
    )
    #%>%
    ## replace NA with the literal string "NA" so str_* functions from stringr can deal with it
    #dplyr::mutate(
    #  remark = stringr::str_replace_na(remark)
    #)

  Licor$data <-
    dat_temp3

  return(Licor)
}


R/detect_Licor_model_version.R
#' Detect the Licor model version
#'
#' @param Licor_fn Licor filename to read
#' @param n_header_rows_min Number of lines to read, should be more than the number of header lines
#'
#' @return Licor_model includes \code{model} number or "Not Detected"
#' @export
#'
detect_Licor_model_version <-
  function(
    Licor_fn
  , n_header_rows_min = 100
  ) {

  ## DEBUG
  # path <- "C:/Dropbox/StatAcumen/consult/Authorship/2009_DavidHanson_Isotopes/R-package/RLicor/data-raw"
  # setwd(path)
  # Licor_fn <- "data_Licor_6400_6-1-3.txt"
  # Licor_fn <- "data_Licor_6400_6-1-4.txt"
  # Licor_fn <- "data_Licor_6800_.txt"

  # list can hold more options later, if helpful
  Licor_model <-
    list()

  # Read Licor file header
  message("Reading first ", n_header_rows_min, " lines of Licor file: ", Licor_fn)

  Licor_head <-
    base::readLines(
      Licor_fn
    , n = n_header_rows_min
    )

  # Detect Licor 6800
  # if (grepl("[Header]", Licor_head, fixed = TRUE)) {
  if (any(grepl("Chamber type\t6800", Licor_head, fixed = TRUE))) {
    Licor_model$model   <- "6800"
    return(Licor_model)
  }

  # Detect Licor 6400
  if (any(grepl("<li6400>", Licor_head, fixed = TRUE))) {
    Licor_model$model   <- "6400"
    return(Licor_model)
  }

  # Otherwise, not detected
  Licor_model$model   <- "Not Detected"
  return(Licor_model)
}



R/read_Licor.R
#' Read Licor 6400 and 6800 files
#'
#' @param Licor_fn Licor file to read
#' @param sw_model Model of Licor or autodetect
#' @param n_header_rows_min Minimum number of rows to read in order to detect where header ends and data begins
#'
#' @return Licor is a list including the header and data
#' @export
#'
read_Licor <-
  function(
    Licor_fn
  , sw_model   = c("autodetect", 6800, 6400)[1]
  , n_header_rows_min = 100
  ) {

  ## DEBUG
  # library(RLicor)
  # path <- "C:/Dropbox/StatAcumen/consult/Authorship/2009_DavidHanson_Isotopes/package_testing/RLicor/data-raw"
  # setwd(path)
  # Licor_fn <- "data_Licor_6400_6-1-3.txt"
  # Licor_fn <- "data_Licor_6400_6-1-4.txt"
  # Licor_fn <- "data_Licor_6800_.txt"
  # dat_Licor <- read_Licor(Licor_fn)
  # str(dat_Licor)

  # list can hold more options later, if helpful
  Licor_model <-
    list()

  # Determine Licor model and version
  if (!any(sw_model %in% c("autodetect", 6800, 6400))) {
    warning("Licor sw_model argument not set correctly, defaulting to \"autodetect\".")
    sw_model <- "autodetect"
  }
  if (sw_model == 6800) {
    Licor_model$model <- 6800
  }
  if (sw_model == 6400) {
    Licor_model$model <- 6400
  }
  if (sw_model == "autodetect") {
    Licor_model <- detect_Licor_model_version(Licor_fn, n_header_rows_min)
  }

  if (Licor_model$model == "Not Detected") {
    stop("Licor model (6800 or 6400) could not be automatically detected from file header.")
  }


  # Read appropriate version

  if (Licor_model$model == 6800) {
    message("Licor 6800 model")
    Licor <- read_Licor_6800(Licor_fn)
  }

  if (Licor_model$model == 6400) {
    message("Licor 6400 model")
    Licor <- read_Licor_6400(Licor_fn)
  }

  return(Licor)
}

